File: src/lib/ai/browser-classify.ts
/**
 * Browser-use risk classification.
 *
 * Mirrors `shell-classify.ts` for the browser tool surface. Used by
 * `inference-with-tools.ts` to decide whether a browser_* call requires
 * user approval via `onConfirmExecution`. The Rust side enforces the same
 * rules in `src-tauri/src/browser_classify.rs` — never trust this
 * classifier alone.
 *
 * Tiers:
 *   - read         : autonomous (observe, navigate to http(s), open, close, extract).
 *   - write        : requires approval (form submits, password typing, click on form_submit).
 *   - destructive  : requires approval, hard-default risky (file:/javascript:/mailto:/tel: navigation).
 */

export type BrowserRisk = 'read' | 'write' | 'destructive';

/** AX node tag set the sidecar emits via snapshot.ts. Mirrored here so the
 *  classifier can reason about browser_act intent before dispatch. */
export interface BrowserActHints {
    /** Tags from the AX node at params.index (set by the latest browser_observe). */
    tags?: string[];
    /** AX role of the node (button / textbox / link / …). */
    role?: string;
}

interface ClassifyArgs {
    method: string; // browser_open / browser_navigate / browser_observe / browser_act / browser_close / browser_extract
    params: Record<string, unknown>;
    /** Hints derived from the latest browser_observe — only meaningful for browser_act. */
    hints?: BrowserActHints;
}

const DESTRUCTIVE_URL_SCHEMES = /^(mailto:|tel:|file:|javascript:)/i;

export function classifyBrowserAction({ method, params, hints }: ClassifyArgs): BrowserRisk {
    switch (method) {
        case 'browser_open':
        case 'browser_close':
        case 'browser_observe':
        case 'browser_extract':
        case 'browser_mark':
            return 'read';

        case 'browser_navigate': {
            const url = String(params.url ?? '').trim();
            if (!url) return 'write'; // default closed on bogus input
            if (DESTRUCTIVE_URL_SCHEMES.test(url)) return 'destructive';
            if (/^https?:\/\//i.test(url)) return 'read';
            return 'write';
        }

        case 'browser_act': {
            const action = String(params.action ?? '').toLowerCase();
            const submit = params.submit === true;
            const tags = hints?.tags ?? [];

            // Explicit form submit → write regardless of action shape.
            if (submit) return 'write';
            // Typing into a password / credit-card field → write.
            if (action === 'type' && tags.includes('password')) return 'write';
            // Clicking a button inside a sensitive form → write.
            if (action === 'click' && tags.includes('form_submit')) return 'write';
            // press Enter on a password-tagged field is a credential submit.
            if (action === 'press' && tags.includes('password')) return 'write';

            // Pure-read interactions: hover, scroll, plain typing into search-like
            // boxes (no password tag), single-target clicks on non-submit elements.
            if (['hover', 'scroll'].includes(action)) return 'read';
            if (action === 'type' || action === 'select' || action === 'press' || action === 'click') {
                return 'read';
            }
            return 'write'; // unknown action — default closed
        }

        default:
            // Unknown browser_* method. Default closed: model has to ask
            // explicit permission to do something we don't recognize.
            return 'write';
    }
}

/**
 * Human-friendly one-line summary of an intended browser action — rendered
 * on confirm_action cards alongside the screenshot ("Click 'Reset password'
 * on https://admin.example.com/users/123").
 */
export function describeBrowserAction(args: ClassifyArgs, url: string | undefined): string {
    const { method, params, hints } = args;
    const u = url ?? (params.url as string | undefined) ?? '';
    switch (method) {
        case 'browser_navigate':
            return `Navigate to ${params.url ?? '(missing)'}`;
        case 'browser_act': {
            const action = String(params.action ?? '');
            const name = hints?.role && (params as Record<string, unknown>).name
                ? `${hints.role} "${(params as Record<string, unknown>).name}"`
                : `element [${params.index ?? '?'}]`;
            const onUrl = u ? ` on ${u}` : '';
            if (action === 'click') return `Click ${name}${onUrl}`;
            if (action === 'type') return `Type into ${name}${onUrl}`;
            if (action === 'press') return `Press "${params.text ?? '?'}" on ${name}${onUrl}`;
            if (action === 'select') return `Select "${params.text ?? '?'}" in ${name}${onUrl}`;
            return `${action} ${name}${onUrl}`;
        }
        default:
            return method;
    }
}

File: src-tauri/src/browser_classify.rs
// Authoritative browser-use risk classification.
//
// Mirrors `src/lib/ai/browser-classify.ts`. Called from `browser_rpc` so the
// Rust side never trusts the frontend's classification alone. Same tiers:
//   - "read"        : autonomous
//   - "write"       : requires approval
//   - "destructive" : requires approval, hard-default risky
//
// browser_act params optionally carry `tags` (from the AX node at
// params.index) and `submit` (boolean). When tags include "password" or
// "form_submit", the action escalates to write.

use serde_json::Value;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BrowserRisk {
    Read,
    Write,
    Destructive,
}

impl BrowserRisk {
    pub fn as_str(&self) -> &'static str {
        match self {
            BrowserRisk::Read => "read",
            BrowserRisk::Write => "write",
            BrowserRisk::Destructive => "destructive",
        }
    }
}

fn url_is_destructive(url: &str) -> bool {
    let trimmed = url.trim_start().to_ascii_lowercase();
    trimmed.starts_with("mailto:")
        || trimmed.starts_with("tel:")
        || trimmed.starts_with("file:")
        || trimmed.starts_with("javascript:")
}

fn has_tag(params: &Value, tag: &str) -> bool {
    params
        .get("tags")
        .and_then(|v| v.as_array())
        .map(|arr| arr.iter().any(|t| t.as_str() == Some(tag)))
        .unwrap_or(false)
}

pub fn classify(method: &str, params: &Value) -> BrowserRisk {
    match method {
        "browser.open" | "browser.close" | "browser.observe" | "browser.extract" | "browser.mark" => BrowserRisk::Read,

        "browser.navigate" => {
            let url = params.get("url").and_then(|v| v.as_str()).unwrap_or("");
            if url.is_empty() {
                return BrowserRisk::Write;
            }
            if url_is_destructive(url) {
                return BrowserRisk::Destructive;
            }
            let lower = url.to_ascii_lowercase();
            if lower.starts_with("http://") || lower.starts_with("https://") {
                BrowserRisk::Read
            } else {
                BrowserRisk::Write
            }
        }

        "browser.act" => {
            let action = params.get("action").and_then(|v| v.as_str()).unwrap_or("").to_ascii_lowercase();
            let submit = params.get("submit").and_then(|v| v.as_bool()).unwrap_or(false);
            if submit {
                return BrowserRisk::Write;
            }
            let password_tagged = has_tag(params, "password");
            let form_submit_tagged = has_tag(params, "form_submit");
            match action.as_str() {
                "type" | "press" if password_tagged => BrowserRisk::Write,
                "click" if form_submit_tagged => BrowserRisk::Write,
                "hover" | "scroll" | "click" | "type" | "select" | "press" => BrowserRisk::Read,
                _ => BrowserRisk::Write, // unknown action — default closed
            }
        }

        _ => BrowserRisk::Write, // unknown method — default closed
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn read_methods_are_read() {
        for m in ["browser.open", "browser.close", "browser.observe", "browser.extract", "browser.mark"] {
            assert_eq!(classify(m, &json!({})), BrowserRisk::Read, "{}", m);
        }
    }

    #[test]
    fn navigate_classifies_by_scheme() {
        assert_eq!(classify("browser.navigate", &json!({ "url": "https://example.com" })), BrowserRisk::Read);
        assert_eq!(classify("browser.navigate", &json!({ "url": "http://example.com" })), BrowserRisk::Read);
        assert_eq!(classify("browser.navigate", &json!({ "url": "mailto:a@b" })), BrowserRisk::Destructive);
        assert_eq!(classify("browser.navigate", &json!({ "url": "tel:+15555550100" })), BrowserRisk::Destructive);
        assert_eq!(classify("browser.navigate", &json!({ "url": "file:///etc/passwd" })), BrowserRisk::Destructive);
        assert_eq!(classify("browser.navigate", &json!({ "url": "javascript:alert(1)" })), BrowserRisk::Destructive);
        assert_eq!(classify("browser.navigate", &json!({ "url": "" })), BrowserRisk::Write);
    }

    #[test]
    fn act_password_tag_escalates_write() {
        assert_eq!(
            classify("browser.act", &json!({ "action": "type", "tags": ["password"] })),
            BrowserRisk::Write,
        );
        assert_eq!(
            classify("browser.act", &json!({ "action": "type", "tags": [] })),
            BrowserRisk::Read,
        );
    }

    #[test]
    fn act_form_submit_tag_escalates_write() {
        assert_eq!(
            classify("browser.act", &json!({ "action": "click", "tags": ["form_submit"] })),
            BrowserRisk::Write,
        );
        assert_eq!(
            classify("browser.act", &json!({ "action": "click" })),
            BrowserRisk::Read,
        );
    }

    #[test]
    fn act_submit_flag_escalates_write() {
        assert_eq!(
            classify("browser.act", &json!({ "action": "type", "submit": true })),
            BrowserRisk::Write,
        );
    }

    #[test]
    fn unknown_method_defaults_write() {
        assert_eq!(classify("browser.weirdthing", &json!({})), BrowserRisk::Write);
    }
}
